### ISSUE TITLE
Netbird not working on Unifi after Firmware 4.0

### ISSUE THREAD

**Issue by Llamrei80** [netbirdio/netbird#2116]
Since the update to beta firmware 4.0 on the Unifi Dream Machine Pro Netbird doesn't connect anymore, it worked fine on firmware 3.2.12.

**Are you using NetBird Cloud?**

Yes

**NetBird version**

0.27.10

**NetBird status -d output:**

OS: linux/arm64
Daemon version: 0.27.10
CLI version: 0.27.10
Management: Connected to https://api.netbird.io:443
Signal: Connected to https://signal.netbird.io:443
Relays: 
  [stun:stun.netbird.io:5555] is Unavailable, reason: dial: failed to listen: dial: dial udp 3.73.3.142:5555: connect: invalid argument
  [turns:turn.netbird.io:443?transport=tcp] is Unavailable, reason: dial: dial: dial tcp 18.157.58.205:443: connect: invalid argument

**Comment by nicdercole** [comment]
I have the same identical problem, but my version is stuck at 25.5. Any version higher than that does not allow me to connect to the Netbird cloud.

**Comment by nicdercole** [comment]
Daemon version: 0.25.5
CLI version: 0.25.5
Management: Connected to https://api.netbird.io:443
Signal: Connected to https://signal.netbird.io:443
Relays: 
  [stun:stun.netbird.io:5555] is Available
  [turns:turn.netbird.io:443?transport=tcp] is Available
FQDN: *******.netbird.cloud
NetBird IP: ***.**.***.*/16
Interface type: Kernel
Peers count: 4/8 Connected


____________________________________


2024-07-09T13:35:29+02:00 WARN client/cmd/root.go:234: retrying Login to the Management service in 1.299560714s due to error rpc error: code = DeadlineExceeded desc = context deadline exceeded
Error: login backoff cycle failed: rpc error: code = DeadlineExceeded desc = context deadline exceeded
root@****:/persistent/dpkg/bullseye/packages# netbird status -d
Daemon status: LoginFailed

Run UP command to log in with SSO (interactive login):

 netbird up 

If you are running a self-hosted version and no SSO provider has been configured in your Management Server,
you can use a setup-key:

 netbird up --management-url <YOUR_MANAGEMENT_URL> --setup-key <YOUR_SETUP_KEY>

More info: https://docs.netbird.io/how-to/register-machines-using-setup-keys

**Comment by Llamrei80** [comment]
Just tried downgrading and testet a few versions:

0.26.3 is the last one that works for me. Everything after that gets the same error as in my first post.

Daemon version: 0.26.3
CLI version: 0.26.3
Management: Connected to https://api.netbird.io:443
Signal: Connected to https://signal.netbird.io:443
Relays:
  [stun:stun.netbird.io:5555] is Available
  [turns:turn.netbird.io:443?transport=tcp] is Available
Nameservers:
  [192.168.42.1:53] for [benjaminbu.ch] is Available
FQDN: dream-machine-pro.netbird.cloud
NetBird IP: 100.66.72.58/16
Interface type: Kernel
Quantum resistance: false
Routes: -
Peers count: 2/5 Connected

**Comment by NightSkySK** [comment]
The same for me at Unifi Cloud Gateway Ultra. Netbird 0.28.6 doesn't work. Following above suggestion after downgrade to 0.26.3 everything start working

**Comment by zhangnew** [comment]
same issue on openwrt, Netbird 0.26.3 works for me

**Comment by nazarewk** [comment]
I'm also affected on RutOS (OpenWRT derivative) @ Teltonika RUTX50

**Comment by nazarewk** [comment]
looks like somebody made a progress with this or similar issue at https://github.com/netbirdio/netbird/issues/2512

**Comment by nazarewk** [comment]
Thanks to somebody's suggestion on Slack I tried using legacy operation mode using [mikrotik guide](https://docs.netbird.io/how-to/client-on-mikrotik-router) and it started working.

basically:
```shell
mkdir /etc/sysconfig
cat <<'EOF' >/etc/sysconfig/netbird
export NB_LOG_LEVEL=info
export NB_DISABLE_CUSTOM_ROUTING=true
export NB_USE_LEGACY_ROUTING=true
EOF
/etc/init.d/netbird start
```

**Comment by nazarewk** [comment]
After some Slack explanation of the legacy mode I have confirmed my router (and probably all other Teltonik/RUTOS routers) does not use `nftables`, still stuck at the legacy `iptables` and this is the reason for the weird error of `invalid argument`.

This could get documented somewhere @mlsmaycon

**Comment by nazarewk** [comment]
FYI: on the Slack we've checked for `fwmark`/`SO_MARK` support on the system using this Python script (there was 3.9.7 in available on RutOS), which appears to be properly supported:

```
root@yelk:~# python netbird-socket-test.py 
Successfully set SO_MARK option
OK
root@yelk:~# cat netbird-socket-test.py 
import socket
import sys

def test_fwmark():
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_MARK, 100)
        
        print("Successfully set SO_MARK option")
        return True
    except OSError as e:
        print(f"Failed to set SO_MARK option: {e}")
        return False
    finally:
        sock.close()

if __name__ == "__main__":
    if test_fwmark():
        print('OK')
        sys.exit(0)
    else:
        print('ERROR')
        sys.exit(1)
```

**Comment by nazarewk** [comment]
FYI: after some [source code patching and rebuilding](https://github.com/nazarewk-iac/nix-configs/commit/b716351243bca4a0a57a38f34a8d939d86984bce), current [conclusion by @lixmal](https://netbirdio.slack.com/archives/C02KHAE8VLZ/p1725399479165009?thread_ts=1725351531.376339&cid=C02KHAE8VLZ) is:
> I see, so there's some unknown issue with *using* the fwmark

**Comment by mlsmaycon** [comment]
Thanks for validating the solution @nazarewk. 

With 0.26.4+, the Linux client uses IP rules and firewall marks to handle routing. The flags NB_DISABLE_CUSTOM_ROUTING and NB_USE_LEGACY_ROUTING disable this behavior and fall back to the previous functionality. The downside of having these enabled is that these devices can't be clients of exit nodes. 

I am almost certain, @lixmal can correct me if I am wrong, that only the NB_USE_LEGACY_ROUTING would be enough for your case.

**Comment by nazarewk** [comment]
> I am almost certain, @lixmal can correct me if I am wrong, that only the NB_USE_LEGACY_ROUTING would be enough for your case.

Yes it did, let's continue on my specific use case at #2530 , added a lot more informatioin there extracted from the Slack thread.

**Comment by saket424** [comment]
any further suggestions on this? the NB_USE_LEGACY_ROUTING does not appear to be working for me with v0.33.0

**Comment by SISheogorath** [comment]
I do run a current netbird version on my UDM using the following scripts:

https://git.shivering-isles.com/-/snippets/22

**Comment by lixmal** [comment]
Can you try with stock NetBird, latest version? There is some auto detection for fwmark/ip rule support

**Comment by kwhelchel** [comment]
IS there anyone that can help me get this setup and running?
Site A is udm pro Site B udr7 I want site to site to have site B ping as site A wan

**Comment by donovan-esterhuizen** [comment]
We have the same problem with Unifi Cloud Gateway running firmware 4.,x

**Comment by Homie13** [comment]
SISheogorath has updated the instruction for using his script. It is actual working with firmware 4.,x ...
Many thanks to SISheogorath. I have a UCG Ultra with UniFi OS 4.3.9 and the script is running perfectly.
Many Thanks to SISheogorath.

**Comment by saket424** [comment]
@Homie13 
While I managed to get netbird running on the ubiquiti dream machine SE device, I am not able to add the wt0 interface into the zone firewall as a valid interface. I can ping the netbird IP address but can't ssh to it. Any idea how to update the firewall rule to allow inbound ssh on the netbird interface?

**Comment by Homie13** [comment]
@saket424 
Login in to your Netbird Account. Go to Peers. Identify the relevant peer. On the right side click the three dots and
click on Enable SSH access. Try this, perhaps this could function.

**Comment by saket424** [comment]
@Homie13 
I am self-hosting the netbird instance. It did not seem to matter whether I enabled ssh in the management portal. I am only able to ping the address but unable to ssh to it because the UDM firewall is probably blocking traffic on the zt0 interface by default. Does inbound ssh over netbird work for you when you are remote? My difficulty is how to add zt0 into a zone that is allowed. I can't seem to add it as a new interface via the zone gui

**Comment by herman-rogers** [comment]
> I do run a current netbird version on my UDM using the following scripts:
> 
> https://git.shivering-isles.com/-/snippets/22

FWIW this work for me on version 0.55.1 of netbird on the UDM.

**Comment by M0nk3yOo** [comment]
Does anyone else has the issue, that netbird can't connect and generate this error (if this is the root cause)

> ERRO client/firewall/iptables/router_linux.go:117: failed to clean up rules from FORWARD chain: clear and delete chain NETBIRD-RT-PRE in table mangle: running [/usr/sbin/iptables -t mangle -X NETBIRD-RT-PRE --wait]: exit status 1: iptables: Too many links.

Some lines later:
> 2026-02-03T11:42:32+01:00 ERRO client/internal/connect.go:316: error while starting Netbird Connection Engine: create firewall manager: init firewall: router init: create containers: create chain NETBIRD-RT-PRE in t                                                                                                                 able mangle: running [/usr/sbin/iptables -t mangle -N NETBIRD-RT-PRE --wait]: exit status 1: iptables: Chain already exists.

I already tried to remove this entries via 

> iptables -F and iptables -X


I installed NetBird with help of the previous mentioned script :-)
The agent also was able to "connect", because the key is used, and the peer is listed in the peers list
My NetBird is selfhosted

**Comment by berry** [comment]
> Does anyone else has the issue, that netbird can't connect and generate this error (if this is the root cause)
> 
> > ERRO client/firewall/iptables/router_linux.go:117: failed to clean up rules from FORWARD chain: clear and delete chain NETBIRD-RT-PRE in table mangle: running [/usr/sbin/iptables -t mangle -X NETBIRD-RT-PRE --wait]: exit status 1: iptables: Too many links.
> 
> Some lines later:
> 
> > 2026-02-03T11:42:32+01:00 ERRO client/internal/connect.go:316: error while starting Netbird Connection Engine: create firewall manager: init firewall: router init: create containers: create chain NETBIRD-RT-PRE in t                                                                                                                 able mangle: running [/usr/sbin/iptables -t mangle -N NETBIRD-RT-PRE --wait]: exit status 1: iptables: Chain already exists.
> 
> I already tried to remove this entries via
> 
> > iptables -F and iptables -X
> 
> I installed NetBird with help of the previous mentioned script :-) The agent also was able to "connect", because the key is used, and the peer is listed in the peers list My NetBird is selfhosted

Yes, I had the same issue yesterday. I also used the previous mentioned script on my UDM. Same errors. With help from Claude (I am not an expert on this subject) this seems to be the problem:

> Issue:
> NetBird fails to start with iptables: Too many links / Chain already exists errors.
> 
> Root cause:
> Kernel 4.19.152 limits iptables subchains to 64 per table. UbiOS already uses 58 chains in filter table. During initialization, NetBird's cleanup process temporarily exceeds this limit, causing it to fail deleting old chains, then fail creating new ones because they still exist - entering an infinite loop.
> 

For now I have removed every trace of Netbird on my UDM.

**Comment by M0nk3yOo** [comment]
Thanks for your reply. Bad news.

Do you have an idea for a workaround? Have you implemented an alternative way? Another network and try to route traffic through another netbird client?

I had the hope to use my public IP, create a port forwarding to a NetBird client.

**Comment by bluewalk** [comment]
Any update regarding this issue? Getting the same errors on Unifi OS 5.1.26. It does work on a cloud gateway max with Unifi OS 5.1.19 which appears to be running a newer kernel (5.4.213-ui-ipq5322) compared to the UDMPro (4.19.152-ui-alpine).

For now I found this and it works:
Adding the following environment variables to the netbird.service file
```
Environment="NB_USE_LEGACY_ROUTING=true"
```
